fdbserver/DataDistribution.actor.cpp

```diff
@@ -1610,15 +1746,70 @@ ACTOR Future<Void> loadAndDispatchAuditRange(Reference<DataDistributor> self,
 	return Void();
 }
 
+ACTOR Future<Void> scheduleAuditForMetadata(Reference<DataDistributor> self, std::shared_ptr<DDAudit> audit) {
+	if (audit->coreState.getType() != AuditType::ValidateLocationMetadata &&
+	    audit->coreState.getType() != AuditType::ValidateStorageServerShard) {
+		throw not_implemented();
+	}
+	TraceEvent(SevInfo, "DDScheduleAuditForMetadataBegin", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("Range", audit->coreState.range)
+	    .detail("AuditType", audit->coreState.getType());
+	try {
+		state ServerWorkerInfos serverWorkers = wait(self->txnProcessor->getServerListAndProcessClasses());
+		if (audit->coreState.getType() == AuditType::ValidateLocationMetadata) {
+			// Randomly pick a storage server to compare KeyServers and ServerKeys
+			const int idx = deterministicRandom()->randomInt(0, serverWorkers.servers.size());
+			StorageServerInterface targetServer = serverWorkers.servers[idx].first;
+			AuditStorageRequest req(
+			    audit->coreState.id, systemKeys, audit->coreState.getType()); // use systemKeys to represent this goal
+			audit->actors.add(doAuditOnStorageServer(self, audit, targetServer, req));
+			wait(delay(0.1));
+		} else if (audit->coreState.getType() == AuditType::ValidateStorageServerShard) {
```

Comment:
Is the progress persisted? If one storage server fails to do the audit or DD restarts, do we need to start over audit for all storage servers?

Reply:
No, the progress is not persisted. I will add this functionality.